src/binary/timestamp.rs

```diff
@@ -1,95 +1,85 @@
 // Copyright Amazon.com, Inc. or its affiliates.
 
-use std::io::{self, Write};
+use std::io::Write;
 
-use bytes::BufMut;
+use arrayvec::ArrayVec;
 use chrono::{Datelike, Timelike};
-use smallvec::SmallVec;
 
 use crate::{
     binary::{uint::DecodedUInt, var_int::VarInt, var_uint::VarUInt, writer::MAX_INLINE_LENGTH},
     result::IonResult,
     types::timestamp::{Precision, Timestamp},
 };
 
+const ENCODED_BUFFER_SIZE: usize = 16;
+
 /// Extends [`Timestamp`] to support [Ion binary].
 ///
 /// [Ion binary]: https://amzn.github.io/ion-docs/docs/binary.html#6-timestamp
-impl Timestamp {
-    const BUFFER_SIZE: usize = 16;
+pub trait TimestampBinaryExt {
+    fn encode_binary(&self) -> IonResult<ArrayVec<u8, { ENCODED_BUFFER_SIZE }>>;
+    fn write_binary<W: Write>(&self, sink: &mut W) -> IonResult<()>;
+}
```

Comment:
There are two things going on in this trait that seem like a strange conflation of responsibilities.

One is the ability to encode a `Timestamp` into some octets, and the other is to write the `Timestamp` with its type header to some `Write` sink.

I might frame this a little differently:

```rust
pub const MAX_TIMESTAMP_LENGTH = 16;

/// Provides the ability to serialize [`Timestamp`] to some kind of sink.
pub trait TimestampBinaryEncoder {
    /// Encodes the content of a [`Timestamp`] as per the Ion binary encoding.
    /// Returns the length of the encoded bytes.
    ///
    /// This does not encode the type descriptor nor the associated length.
    /// Prefer [`TimestampBinaryEncoder::encode_timestamp_value`] for that.
    fn encode_timestamp(&mut self, timestamp: &Timestamp) -> IonResult<usize>;

    /// Encodes a [`Timestamp`] as an Ion value with the type descriptor and length.
    /// Returns the length of the encoded bytes.
    fn encode_timestamp_value(&mut self, timestamp: &Timestamp) -> IonResult<usize>;
}

impl<T> TimestampBinaryEncoder for T where T: Write {
    ...
}
```

What this factoring gives us is the ability to write to a buffer or sink of our choosing without requiring an intermediate:

```rust
let mut sink: ArrayVec<u8, MAX_TIMESTAMP_LENGTH> = ArrayVec::new_const();
sink.encode_timestamp(&some_ts);
...
```

I realize the intermediate is necessary for the writing of the full value, but that is in some ways an implementation detail and this API hides that.

This also will work nicely as a pattern for the other types we'll expose, so we can model them each as their own traits and pull them together via a trait that is constrained by all the typed encoders and blankets over any type that implements all the specific encoders which all `Write` types will.

Note also that arbitrary precision timestamps will probably make `MAX_TIMESTAMP_LENGTH` into something like `TYPICAL_TIMESTAMP_LENGTH` and necessitate the use of `SmallVec` or some such, but that is an exercise for another day.

Reply:
Very nice. Thanks. I made those changes.

Note that the usize return isn't yet possible due to other changes that need to happen. For now, I left a TODO (in this PR). If we can settle on these APIs, I will also open an issue to track that.